### PR TITLE
Rework internal API

### PR DESCRIPTION
**Pull request by achouippe**
- Add an OpenAPI descriptor for the internal API,
- Support one or multiple topics in one HTTP call,
- Support one or multiple published message in one HTTP cal,
- Validation of the JSON input,
- Split existing unit tests + comprehensive unit tests on the validation logic,
- Possibility to pass message timestamps in the publish request.

### CODE REVIEW

File: neurow/lib/neurow/internal_api/endpoint.ex [hunk 1]
```diff
@@ -53,21 +56,31 @@ defmodule Neurow.InternalApi do
...
+            Phoenix.PubSub.broadcast!(
+              Neurow.PubSub,
+              topic,
+              {:pubsub_message,
+               %Message{message | timestamp: message.timestamp || publish_timestamp}}
+            )
 
-        Logger.debug("Message published on topic: #{topic}")
-        Stats.inc_msg_received()
+            Stats.inc_msg_received()
```
Comment (bpaquet) [inline neurow/lib/neurow/internal_api/endpoint.ex:73]: Not sure what we really want to monitor with the new API.

File: neurow/lib/neurow/internal_api/endpoint.ex [hunk 2]
```diff
@@ -53,21 +56,31 @@ defmodule Neurow.InternalApi do
...
+            Logger.debug("Message published on topic: #{topic}")
+          end)
+        end)
 
         conn
-        |> put_resp_header("content-type", "text/html")
-        |> send_resp(200, "Published #{message} to #{topic}\n")
+        |> put_resp_header("content-type", "application/json")
+        |> send_resp(200, "{\"nb_published\": #{nb_publish}}")
```
Comment (bpaquet) [inline neurow/lib/neurow/internal_api/endpoint.ex:80]: Jason.encode?

File: neurow/lib/neurow/internal_api/message.ex [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+  def from_json(payload) when is_map(payload) do
+    %Neurow.InternalApi.Message{
+      type: payload["type"],
+      timestamp: payload["timestamp"],
+      payload: payload["payload"]
+    }
+  end
+
+  def validate(message) do
```
Comment (bpaquet) [inline neurow/lib/neurow/internal_api/message.ex:19]: You can probably write this with a list of guard. May be it's cleaner.

File: neurow/lib/neurow/internal_api/endpoint.ex (hunk 1 above)
Comment (bpaquet) [reply neurow/lib/neurow/internal_api/endpoint.ex:73]: Should we incr with nb_publish?

File: neurow/lib/neurow/internal_api/publish_request.ex [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+      :ok <- validate_topics(request),
+      :ok <- validate_messages(request)
+    ) do
+      :ok
+    else
+      error -> error
+    end
+  end
+
+  def validate_topics(request) do
```
Comment (bpaquet) [inline neurow/lib/neurow/internal_api/publish_request.ex:52]: Same here, with guards?

File: neurow/lib/neurow/internal_api/publish_request.ex [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            true -> :ok
+            false -> {:error, "'topics' must be an non-empty array of non-empty strings"}
+          end
+        else
+          {:error, "'topics' must be an non-empty array of non-empty strings"}
+        end
+    end
+  end
+
+  def validate_messages(request) do
```
Comment (bpaquet) [inline neurow/lib/neurow/internal_api/publish_request.ex:77]: Ditto

File: neurow/lib/neurow/internal_api/publish_request.ex [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    %Neurow.InternalApi.PublishRequest{
+      topic: payload["topic"],
+      topics: payload["topics"],
+      message:
+        case payload["message"] do
+          message when is_map(message) ->
+            Neurow.InternalApi.Message.from_json(payload["message"])
+
+          _ ->
+            nil
```
Comment (bpaquet) [inline neurow/lib/neurow/internal_api/publish_request.ex:23]: nil -> nil, but _ -> error?

File: neurow/lib/neurow/internal_api/endpoint.ex (hunk 1 above)
Comment (achouippe) [reply neurow/lib/neurow/internal_api/endpoint.ex:73]: Do we want to count the number of call on the internal interface per issuer ?

File: neurow/lib/neurow/internal_api/message.ex (hunk 3 above)
Comment (achouippe) [reply neurow/lib/neurow/internal_api/message.ex:19]: I tried first, but the combinatory effect make it much difficult to read.

File: neurow/lib/neurow/internal_api/endpoint.ex [hunk 7]
```diff
@@ -53,7 +56,15 @@ defmodule Neurow.InternalApi do
 
   get "/history/:topic" do
     history = Neurow.ReceiverShardManager.get_history(topic)
-    history = Enum.map(history, fn {_, {id, message}} -> %{id: id, message: message} end)
+
+    history =
+      Enum.map(history, fn {_, message} ->
```
Comment (bpaquet) [inline neurow/lib/neurow/internal_api/endpoint.ex:61]: @achouippe Any better solution to serialize?

File: neurow/lib/neurow/internal_api/endpoint.ex (hunk 1 above)
Comment (bpaquet) [reply neurow/lib/neurow/internal_api/endpoint.ex:90]: In that case, it should be outside of the loop :)

**Review by bpaquet — APPROVED** [review]

File: neurow/lib/neurow/internal_api/endpoint.ex (hunk 7 above)
Comment (achouippe) [reply neurow/lib/neurow/internal_api/endpoint.ex:61]: @bpaquet , I replaced with

```
    history =
      Enum.map(history, fn {_, message} ->
        Map.from_struct(message)
      end)
```

File: neurow/lib/neurow/internal_api/publish_request.ex (hunk 6 above)
Comment (achouippe) [reply neurow/lib/neurow/internal_api/publish_request.ex:23]: The validation is performed later, so I prefer not to return an error at this step.

**Pull request merged**